### ISSUE TITLE
Make compiletest look for exact matches in the arch table

### ISSUE THREAD

**Issue by rkruppe** [rust-lang/rust#48893]
#47737 reported a problem with architecture names that are substrings of each other, but the problem is more general than powerpc. #48732 removed the `powerpc` annotations, but [I argued afterwards](https://github.com/rust-lang/rust/issues/47737#issuecomment-370232447), it would be better to improve compiletest such that it doesn't consider e.g. `powerpc` to be a `powerpc64` target. In light of the arguments linked earlier and the additional fact that this *keeps happening* -- #48875 introduces the same problem for MIPS -- I really think we should revert #48732 and fix compiletest instead.

cc @nagisa @jcowgill @debris @alexcrichton

**Comment by alexcrichton** [comment]
Sure! Just needs someone to do the work :)

**Comment by jcowgill** [comment]
> #48875 introduces the same problem for MIPS

I don't think so. The change in #48875 works because "mips64" is inserted above "mips" in the `ARCH_TABLE` list.

**Comment by rkruppe** [comment]
Hm, I think you're right. The correctness of the code depending on such a subtlety is still pretty awful, though.
